filter_throttle_patch/throttle.c

```diff
@@ -0,0 +1,316 @@
+          || ! apply_suffix (&s, *p))
+        {
+            flb_plg_warn(ctx->ins,
+                         "invalid time interval %s falling back to default: 1 "
+                         "second",
+                         interval);
+        }
+
+      seconds += s;
+      return seconds;
+}
+
+static int cb_throttle_init(struct flb_filter_instance *f_ins,
+                        struct flb_config *config,
+                        void *data)
+{
+    int ret;
+    pthread_t tid;
+    struct ticker *ticker_ctx;
+    struct flb_filter_throttle_ctx *ctx;
+
+    /* Create context */
+    ctx = flb_malloc(sizeof(struct flb_filter_throttle_ctx));
+    if (!ctx) {
+        flb_errno();
+        return -1;
+    }
+    ctx->ins = f_ins;
+
+    /* parse plugin configuration  */
+    ret = configure(ctx, f_ins);
+    if (ret == -1) {
+        flb_free(ctx);
+        return -1;
+    }
+
+    ticker_ctx = flb_malloc(sizeof(struct ticker));
+    if (!ticker_ctx) {
+        flb_errno();
+        flb_free(ctx);
+        return -1;
+    }
+
+    /* Set our context */
+    flb_filter_set_context(f_ins, ctx);
+
+    ctx->hash = window_create(ctx->window_size);
+
+    ticker_ctx->ctx = ctx;
+    ticker_ctx->done = false;
+    ticker_ctx->seconds = parse_duration(ctx, ctx->slide_interval);
+    pthread_create(&tid, NULL, &time_ticker, ticker_ctx);
+    return 0;
+}
+
+static int cb_throttle_filter(const void *data, size_t bytes,
+                              const char *tag, int tag_len,
+                              void **out_buf, size_t *out_size,
+                              struct flb_filter_instance *f_ins,
+                              void *context,
+                              struct flb_config *config)
+{
+    int ret;
+    bool retain = false;
+    int old_size = 0;
+    int new_size = 0;
+    msgpack_unpacked result;
+    msgpack_object root;
+    size_t off = 0;
+    (void) f_ins;
+    (void) config;
+    msgpack_sbuffer tmp_sbuf;
+    msgpack_packer tmp_pck;
+
+    /* Create temporary msgpack buffer */
+    msgpack_sbuffer_init(&tmp_sbuf);
+    msgpack_packer_init(&tmp_pck, &tmp_sbuf, msgpack_sbuffer_write);
+
+
+    /* Iterate each item array and apply rules */
+    msgpack_unpacked_init(&result);
+    while (msgpack_unpack_next(&result, data, bytes, &off) == MSGPACK_UNPACK_SUCCESS) {
+        root = result.data;
+        if (root.type != MSGPACK_OBJECT_ARRAY) {
+            continue;
+        }
+
+        old_size++;
+
+        ret = throttle_data(context);
+        retain = get_retain(context);
+        if (ret == THROTTLE_RET_KEEP) {
+            msgpack_pack_object(&tmp_pck, root);
+            new_size++;
+        }
+        else if (ret == THROTTLE_RET_DROP) {
+            /* If Retain is false, Do nothing */
+            if (retain) {
+                usleep(10 * 1000);
+                msgpack_pack_object(&tmp_pck, root);
+                new_size++;
+             }
+        }
```

Comment:
```suggestion
       if (ret == THROTTLE_RET_KEEP) || retain {
            msgpack_pack_object(&tmp_pck, root);
            new_size++;
        }
```

Reply:
why sleep if retain?

Reply:
Could ret have values other than THROTTLE_RET_KEEP and THROTTLE_RET_DROP?

Reply:
1. Sleep for break data stream in fluent-bit framwork, if no sleep, data stream may combine data, result throttle will not work as expected.
2. ret only THROTTLE_RET_KEEP and THROTTLE_RET_DROP, if  ret== THROTTLE_RET_DROP,  drop or not will decide by retain.